File: Python/Math-Array-String/38.py
# 704. Binary Search

from typing import List

class Solution:
    def search(self, nums: List[int], target: int) -> int:
        left, right = 0, len(nums) - 1  # Initialize the search range
    
        while left <= right:
            mid = (left + right) // 2  # Find the middle index
            if nums[mid] == target:   # If the middle element is the target
                return mid
            elif nums[mid] < target: # Target lies in the right half
                left = mid + 1
            else:                    # Target lies in the left half
                right = mid - 1
        
        return -1  # If target is not found


sol = Solution()
output = sol.search([-1,0,3,5,9,12], 9)
print(output)

File: JavaScript/Math-Array-String/38.js
// 704. Binary Search

/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number}
 */
var search = function(nums, target) {
    let left = 0, right = nums.length - 1;

    while (left <= right) {
        let mid = Math.floor((left + right) / 2); // Calculate the middle index

        if (nums[mid] === target) {
            return mid; // Target found, return the index
        } else if (nums[mid] < target) {
            left = mid + 1; // Narrow the search to the right half
        } else {
            right = mid - 1; // Narrow the search to the left half
        }
    }

    return -1; // Target not found
};


console.log(search([-1,0,3,5,9,12], 9));
